### PR TITLE
Unicast DNS Service Discovery

### PR DESCRIPTION
**Pull request by cawhitworth**
# Description

Add an example project for DNS service discovery using unicast DNS.

## Type of change

- New content

# Checklist:

- [X] My content has a file named README.md, which is based on the appropriate readme [template](https://github.com/Azure/azure-sphere-gallery/tree/main/Templates)
- [X] I have performed a self-review of my own contribution
- [X] I have provided comments where appropriate
- [x] I have updated the repository's [README.md](https://github.com/Azure/azure-sphere-gallery/blob/main/README.md) to index this project
- [X] I have added/updated license(s) for this project as appropriate
- [X] I have permission to publish this content (in case the content was collaborative)
- [X] I have verified that my content does not contain sensitive information (PII, Microsoft PI, etc.) and is safe to open source

### CODE REVIEW

**Review by jamesadevine — APPROVED** [review]

File: DNSServiceDiscoveryUnicast/README.md [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+1. Set up a DNS service. A container is provided that runs a minimal `bind9` on Ubuntu 22.04:
+
+   ```console
+   $ docker build ./container -t dns-sd-example
+   $ docker run --rm -it -p 53:53/udp 5353:5353/udp dns-sd-example
+   * Starting domain name service... named                  [ OK ] 
+   root@ef71614eec1c:/#
+   ```
+
+   Note that if you have an existing DNS server running on port 53 of the Docker host (for example, a Linux system such as Ububtu that uses systemd is likely to have systemd-resolved running for local name resolution) you will need to disable this and configure an alternative (for example, specify a nameserver directly in `/etc/resolv.conf`)
```
Comment (CoderScribe) [inline DNSServiceDiscoveryUnicast/README.md:48]: Spell check for Ububtu

File: DNSServiceDiscoveryUnicast/README.md [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+- (optional) a machine running Docker to run the DNS server container
+
+## Setup
+
+1. Even if you've performed this set up previously, ensure that you have Azure Sphere SDK version 22.11 or above. At the command prompt, run **azsphere show-version** to check. Install [the Azure Sphere SDK](https://learn.microsoft.com/azure-sphere/install/install-sdk) as needed.
+1. Connect your Azure Sphere device to your computer by USB.
+1. Connect your Azure Sphere device to the same local network as the DNS service.
+1. Enable application development, if you have not already done so, by entering the **azsphere device enable-development** command at the command prompt.
+
+1. Set up a DNS service. A container is provided that runs a minimal `bind9` on Ubuntu 22.04:
```
Comment (CoderScribe) [inline DNSServiceDiscoveryUnicast/README.md:39]: There's an assumption that the user knows how to discover the IP address of their docker container, is there something we can point them to?

File: DNSServiceDiscoveryUnicast/README.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+```
+ERROR: Invalid DNS server address or address family specified.
+```
+
+ensure you have configured the DNS servers as described above.
+
+## Key concepts
+
+Unlike multicast service discovery, the discovered service may exist outside of the local network for the device. This means that DNS service discovery can be used to discovery and connect to endpoints on the wider internet - for example, Azure IoT hubs or other dynamically-assigned endpoints - without needing to specify them in the application manifest.
```
Comment (CoderScribe) [inline DNSServiceDiscoveryUnicast/README.md:105]: spell check: discovery == discover

File: DNSServiceDiscoveryUnicast/README.md (hunk 2 above)
Comment (cawhitworth) [reply DNSServiceDiscoveryUnicast/README.md:39]: The `docker run` command binds the container's DNS ports to the host's, so the IP address of the docker container doesn't matter - it's the IP address of the host running the container that will appear as the DNS server. I'll clarify this.

**Pull request merged**